---
title: "TCGA data download"
author: "Antonio Ahn"
date: "`r Sys.Date()`"
output:
<<<<<<< HEAD
  pdf_document:
    toc: yes
  html_document:
    df_print: paged
    toc: yes
=======
  html_document:
    df_print: paged
    toc: yes
  pdf_document:
    toc: yes
>>>>>>> 4203343b6214a072c3feb85611f00ef04d28d99e
  rmarkdown::html_vignette:
    number_sections: yes
    toc: yes
references:
- author:
  - family: Samur, M. K.
    given: null
  id: ref1
  issued:
    year: 2014
  journal: PLoS One
  number: 9
  pages: null
  title: RTCGAToolbox a new tool for exporting TCGA Firehose data.
  volume: 9
- author:
  - family: Ekmekcioglu, S., et al
    given: null
  id: ref2
  issued:
    year: 2016
  journal: Clin Cancer Res
  number: 12
  pages: 3016-3024
  title: Inflammatory Marker Testing Identifies CD74 Expression in Melanoma Tumor
    Cells, and Its Expression Associates with Favorable Survival for Stage III Melanoma
  volume: 22
- author:
  - family: Jeschke, J., et al.
    given: null
  id: ref3
  issued:
    year: 2017
  journal: J Clin Invest
  number: 8
  pages: 3090-3102
  title: DNA methylation-based immune response signature improves patient diagnosis
    in multiple cancers
  volume: 127
---

#Introduction

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=TRUE)
```

This document outlines the approach i took to download data from level 3 and 4 TCGA data repositry. The clinical information, RNA-seq data and methylation 450K data was downloaded using the `RTCGAToolbox`[@ref1] package and then processed or cleaned-up for downstream analysis.

The `RTCGAToolbox` package retreives data from the broads institutive firehose database. Detailed information on how to use this package and the included functions are available in the [Vignette](https://www.bioconductor.org/packages/devel/bioc/vignettes/RTCGAToolbox/inst/doc/RTCGAToolbox-vignette.html), online [PH525Xseries](http://genomicsclass.github.io/book/pages/tcga.html) courses and [Youtube](http://rafalab.github.io/pages/harvardx.html) videos. 

Load the library
```{r message=FALSE}
library(RTCGAToolbox)
```

A list of available cancer types are found using the `getFirehoseDatasets` command. The run dates and analyze datas are found using the `getFirehoseRunningDates` and the `getFirehoseAnalyzeDates` commands. 

```{r}
getFirehoseDatasets()
head(getFirehoseRunningDates())
head(getFirehoseAnalyzeDates())
```

 
#Clinical and RNA-seq data - download and processing

The clinical information and normalised RNA-seq data is downloaded.
The methylation data is downloaded separately from the clinical and RNA-seq data because its size was too large for my computer to handle. Thus the methylation data was downloaded in the DSM3735 server.

Skin cutaneous melanoma (SKCM) is selected and "20151101" is used as the rundate. The default file size is 500 mb and this limit is extended using `fileSizeLimit`.

```{r cache=TRUE, message=FALSE}
readDataMel <- getFirehoseData (dataset="SKCM", runDate="20151101",forceDownload = TRUE,
<<<<<<< HEAD
    clinical=TRUE, RNASeq2GeneNorm=TRUE, fileSizeLimit= 3000)
=======
    Clinic=TRUE, RNAseq2_Gene_Norm=TRUE, fileSizeLimit= 3000)
>>>>>>> 4203343b6214a072c3feb85611f00ef04d28d99e
```

The RNA-seq data (RNAseq2_Gene_Norm) contains gene expression levels generated using **MapSplice** for alignment and **RNA-Seq by Expectation-Maximization (RSEM)** for quantification. RSEM values are calculated using an algorithm that estimate abundances at the gene level to generate TPM (Transcripts Per Million) values. TPM is similar to FPKM and RPKM in that it accounts for multiple variables including library size and gene length. For normalisation, TPM values are divided by the 75th percentile (3rd quartile) and multiplied by 1000.

##Extracting the data 

Extract the clinical and RNA-seq data.
```{r cache=TRUE, message=FALSE}
<<<<<<< HEAD
clinMel <- getData(readDataMel, "clinical")
=======
clinMel <- getData(readDataMel, "Clinical")
>>>>>>> 4203343b6214a072c3feb85611f00ef04d28d99e
rnaseqMel <- getData(readDataMel, "RNASeq2GeneNorm")
```

#Data cleaning of clinical and RNA-seq information 

The identifiers are structured differently between the clinical and RNA-seq data. The identifiers in the RNA-seq data are transformed to be the same as the ones in the clinical data. Duplicate RNA-seq data are removed and any RNA-seq without clinical information or any clinical information without RNA-seq data are removed. 

*Here i need to add in what the TCGA name means. For example, TCGA-3N-A9WB-06A-11R-A38C-07, what does each section mean. Is there a difference in the raw identifier names between the duplicates?*

##Changing patient identifier names

The identifiers are structure differently between the clinical and RNA-seq data.
```{r}
dim(clinMel)
head(clinMel)

dim(rnaseqMel)
rnaseqMel[1:5,1:5]
```

The identifiers in the RNA-seq data are transformed to be the same as the ones in the clinical data.
```{r cache=TRUE}
rid = tolower(substr(colnames(rnaseqMel),1,12))
rid = gsub("-", ".", rid)  

table(rid %in% rownames(clinMel)) #all 473 RNA-seqMel samples have corresponding clinical details 

length(intersect(rid,rownames(clinMel))) 
# 469 patients out of 470 have RNA-seq data

colnames(rnaseqMel) = rid 
head(colnames(rnaseqMel))
```

##Remove duplicated samples

Samples with duplicated names are removed. These are samples that have 2 RNA-seq data for some reason. The data between the replicates are very similar and thus we remove the second duplicate. 
```{r}
duplicatedSamples <- which(duplicated(colnames(rnaseqMel))) # 4  duplicate samples

duplicatedSampleNames<-colnames(rnaseqMel)[duplicated(colnames(rnaseqMel))]

rnaseqMel_duplicated <-rnaseqMel[,colnames(rnaseqMel) %in% duplicatedSampleNames]

colnames(rnaseqMel_duplicated)

par(mfrow=c(2,2))
plot(log2(rnaseqMel_duplicated[1001:2000,1:2]))
plot(log2(rnaseqMel_duplicated[1001:2000,3:4]))
plot(log2(rnaseqMel_duplicated[1001:2000,5:6]))
plot(log2(rnaseqMel_duplicated[1001:2000,7:8]))
#  it is not obvious which of the duplicates to keep, so we drop the second
```


```{r cache=TRUE}
rnaseqMel = rnaseqMel[,-which(duplicated(colnames(rnaseqMel)))] # getting rid of the duplicate

dim(rnaseqMel) # from 473 samples to 469

length(intersect(colnames(rnaseqMel),rownames(clinMel))) 
length(rownames(clinMel)) # there is 1 sample in clinMel which there is absent in rnaseqMel 

clinMel <-clinMel[intersect(colnames(rnaseqMel),rownames(clinMel)),]
dim(clinMel)

table(colnames(rnaseqMel)==rownames(clinMel)) # patient names are in the same order
```

#Create an expression set

An expression set is created to store the log2 transformed RNA-seq data and the clinical information. 
```{r message=FALSE}
library(Biobase)
```

```{r cache= TRUE}
readES = ExpressionSet(as.matrix(log2(rnaseqMel+1)))
readES
 
exprs(readES)[1:3,1:3]

pData(readES) = clinMel
```


# Survival data clean-up and analysis

## Survival analysis: background

Important pData information for survival analysis is "vital_status", "days_to_death" and "days_to_last_followup"

Information from [here](https://groups.google.com/forum/#!topic/ucsc-cancer-genomics-browser/YvKnWZSsw1Q) 

To analyse overall survival, 3 variables in the `clinMel` data set is required, which are "vital_status", "days_to_death" and "days_to_last_followup".

```{r}
dim(clinMel)
str(clinMel[,c("vital_status","days_to_death","days_to_last_followup")])
clinMel[1:5,c("vital_status","days_to_death","days_to_last_followup")]
```

* **vital_status:** "1" means deceased and "0" means still alive. 
* **days_to_death:** With patients who are deacesed, the days_to_death variable gives the number of days before death.
* **days_to_last_followup:** With patients who are still alive, the days_to_last_followup variable gives the number of days before the last follow-up.


## Survival data: Exploratory analysis

For most patients (391 patients), the days_to_death and days_to_last_followup are mutually exlcusive; if theres an NA in days_to_death then there is a number to DaystoLastfollowup and vice versa.
```{r}
table(!is.na(clinMel[,"days_to_death"]) & is.na(clinMel[,"days_to_last_followup"]))

table(is.na(clinMel[,"days_to_death"]) & !is.na(clinMel[,"days_to_last_followup"]))
```

However there are some patients with both days_to_last_followup and days_to_death (69 patients). Also there are patients with both of these variables as NA (9 patients).
```{r}
survivalVariables <- c("days_to_last_followup","vital_status","days_to_death")

index <- !is.na(clinMel$"days_to_death") & !is.na(clinMel$"days_to_last_followup")
clinMel[index,survivalVariables]
dim(clinMel[index,survivalVariables])

index <- is.na(clinMel[,"days_to_death"]) & is.na(clinMel[,"days_to_last_followup"])
clinMel[index,survivalVariables]
dim(clinMel[index,survivalVariables])
```

There are also some patients with a negative days_to_last_followup. What does this mean?
```{r}
survivalVariables <- c("days_to_last_followup","vital_status","days_to_death")

index <- which(clinMel[,"days_to_death"] < 0 | clinMel[,"days_to_last_followup"] < 0)

clinMel[index,survivalVariables]
```



## Survival analysis: merge days_to_death and days_to_last_followup

Here i merge days_to_death and days_to_last_followup to create a new variable called new_death. Most are simple to handle because they are mutually exlcusive; if there's an NA in days_to_death then there is a number to days_to_last_followup and vice versa. However, as shown above, some patients have values to both variables with different number of days which i am unsure what that means. Also some patients have an NA to both variables. 

Here i create a new variable called new_death.

* If patient has deceased (1 in vital status), the days_to_death is selected
* If patient is alive (0 in vital status), days_to_last_followup is selected

```{r cache=TRUE}
mergeOS <- ifelse(clinMel[,"vital_status"]==1, clinMel[,"days_to_death"], clinMel[,"days_to_last_followup"] )

summary(mergeOS)

clinMel$mergeOS <- as.numeric(mergeOS)
```

clinMel with the `mergeOS` parameter is re-loaded into `readES`.
```{r}
pData(readES) = clinMel
```


## Survival analysis: sanity check with t-stage

* t0 - patients without a known primary tumor.
* t1
* t2
* t3
* t4
* ti
* tx

```{r}
library(survival)
```

```{r}
ev <- as.numeric(pData(readES)$vital_status)
fut <-as.numeric(pData(readES)$mergeOS)
su = Surv(fut, ev)

table(pData(readES)$pathology_T_stage)
table(substr(clinMel$pathology_T_stage,1,2))

t_stage = factor(substr(clinMel$pathology_T_stage,1,2))

plot(survfit(su~t_stage),mark.time=TRUE, lwd=2, col=1:7, las=1, cex.axis=1.5)
mtext("Overall Survival", side=2, line=2.7, cex=1.5)
mtext("Time (Days)", side=1, line=2.8, cex=1.5)

ntab = table(t_stage)
ns = paste("[n=", ntab, "]", sep="")
legend(8000, .8, col=1:7, lwd=2, legend=paste(levels(t_stage), ns))
text(6000,1, paste("p.value = 3.25e-06 "))

summary(coxph(su~t_stage))
survdiff(su~t_stage)
```


There is a significant statistical difference in overall survival between the different T stages


## Survival analysis: sanity check with CD74

CD74 gene exprresion was found to be associated with good prognosis using TCGA data [@ref2]. 

```{r}
CD74 <- ifelse(exprs(readES)["CD74",] > median(exprs(readES)["CD74",]), 1, 0)
# higher than median is 1, lower than median is 0
CD74 <- as.factor(CD74)
table(CD74)

ev <- as.numeric(pData(readES)$vital_status)
fut <-as.numeric(pData(readES)$mergeOS)
su = Surv(fut, ev)

plot(survfit(su~CD74),mark.time=TRUE, lwd=2, col=c("black","red"), las=1, cex.axis=1.5)
mtext("Overall Survival", side=2, line=2.7, cex=1.5)
mtext("Time (Days)", side=1, line=2.8, cex=1.5)

ntab = table(CD74)
ns = paste("[n=", ntab, "]", sep="")
legend(8000, .8, col= c("black","red"), lwd=2, legend=paste(levels(CD74), ns))
text(4000,0.8, paste("p.value= 9.61e-07"))

survdiff(su~CD74, data=clinMel)
```

Higher CD74 gene expression is associated with a better prognosis.


# Methylation 450K data - download and processing

The methylation 450K data-frame was too big (>6gb) to download or work with in my desktop (my dekstop freezes). It has 485,577 rows and 478 columns with each value having many digits. Therefore I had to use the DSM3735 server (based in the pathology department in Otago university) to download the data and then reduce the file size by lowering the number of decimal points for every beta-value. The size-reduced file was then moved to my desktop and loaded into `R`. 

```{r engine='bash', eval=FALSE}
ssh -X aahn@dsm3735.otago.ac.nz # to login to the server

scp aahn@dsm3735.otago.ac.nz:/home/aahn/PDL1/TCGAMel1.RData /Users/antonioahn/Desktop # Move the RData file from the DSM3735 server to my desktop 
```

It is recommended to lower the worker/core usage to prevent crashing the server. 
```{r message=FALSE, eval=FALSE}
library(BiocParallel)
registered()
register(MulticoreParam(workers=2)) 
```

The methylation data are the Beta-values from the 450K methylation arrays. In the DSM3735 server, the methylation data was downloaded, extracted and reduced in size.
```{r eval=FALSE}
readDataMel <- getFirehoseData (dataset="SKCM", runDate="20151101",forceDownload = TRUE,
    Clinic=FALSE, RNAseq2_Gene_Norm=FALSE,Methylation = TRUE, fileSizeLimit= 3000)

me450kMel = getData(readDataMel, "Methylation",1) 
```


```{r eval=FALSE}
dim(me450kMel)
head(colnames(me450kMel))

probeinfo <- me450kMel[,1:3] # These colunms have the "Gene_Symbol", "Chromosome" and "Genomic_Coordinate" information.

write.csv(probeinfo , "probeinfo.csv")
```

```{r engine='bash', eval=FALSE}
scp aahn@dsm3735.otago.ac.nz:/home/aahn/probeinfo.csv /Users/antonioahn/Desktop
```

```{r eval=FALSE }
me450kMel <- me450kMel[,4:478] # dropping the first 3 columns which contains the probe info
```

##Change the identifier names in the methylation data

```{r eval=FALSE}
rid = tolower(substr(colnames(me450kMel),1,12))
rid = gsub("-", ".", rid)

colnames(me450kMel) <- rid

which(duplicated(colnames(me450kMel)))
 
me450kMel <- me450kMel[,!duplicated(colnames(me450kMel))] # dropping the second duplicate samples
dim(me450kMel)


table(duplicated(colnames(me450kMel))) 
# me450kMel has 470 samples but rnaseqMel has 469. There is 1 extra sample in me450kMel.

table(colnames(me450kMel)%in%colnames(rnaseqMel)) 

me450kMel <- me450kMel[,colnames(me450kMel)%in%colnames(rnaseqMel)] # keeping only the matching samples

table(colnames(rnaseqMel)==colnames(me450kMel)) # Everything is in the same length and order. 
 
table(rownames(clinMel)==colnames(me450kMel))
```

##Reducing the size of the methylation 450K data

```{r eval=FALSE}
str(me450kMel) # this shows that all the values are characters. 

me450kMel <- sapply(me450kMel, as.numeric)
me450kMel_rounded <- as.matrix(round(me450kMel, digits=3)) # Round to 3 digits

save.image("/home/aahn/Bioinformatics/RDatafiles/TCGAmelanoma_methylation.RData")
```

After i reduced the size of the methylation data to generate `me450kMel_rounded`, I saved into my computer for loading. 
```{r}
<<<<<<< HEAD
load("~/Dropbox/Education/Bioinformatics/5DataAnalysis/TCGAmelanoma/Methylation/TCGAmelanoma_methylation.RData") # this contains me450kMel_rounded
=======
load("~/Dropbox/Education/Bioinformatics/5. DataAnalysis/TCGAmelanoma/Methylation/TCGAmelanoma_methylation.RData") # this contains me450kMel_rounded
>>>>>>> 4203343b6214a072c3feb85611f00ef04d28d99e

dim(me450kMel_rounded)
dim(probeinfo)
class(me450kMel_rounded)

me450kMel_rounded[1:3,1:3]
```

##Acquring methylation probe values for meTIL-score

It was demonstrated that methylation probe values can be used to determine the level of CD8 immune cells within bulk tumour [@ref3]. 

Beta-values of 5 CpG probes are needed to generate the meTIL-score. Here i did not use me450kMel_rounded but used the data prior to rounding to 3 decimal points.

```{r eval=FALSE}
meTIL_probes <- c("cg20792833","cg20425130","cg23642747","cg12069309","cg21554552") # the 5 CpG probes needed to generate the meTIL-score.

me450kMel[1:3,1:6] 
```
            X Gene_Symbol Chromosome Genomic_Coordinate
1 cg00000029        RBL2         16           53468112
2 cg00000108     C3orf35          3           37459206
3 cg00000109      FNDC3B          3          171916037
  TCGA.3N.A9WB.06A.11D.A38H.05 TCGA.3N.A9WC.06A.11D.A38H.05
1                       0.5167                       0.4193
2                           NA                           NA
3                           NA                           NA
```{r eval=FALSE}
write.csv(me450kMel[me450kMel$X%in%probes_iwant,], file="meTIL_probes.csv")
```

The "meTIL_probes.csv" file is transfered from the server to my computer and then loaded. 

```{r cache=TRUE}
<<<<<<< HEAD
meTIL_probes <- read.csv("~/Dropbox/Education/Bioinformatics/5DataAnalysis/TCGAmelanoma/Methylation/meTIL_probes.csv", row.name=1)
=======
meTIL_probes <- read.csv("~/Dropbox/Education/Bioinformatics/5. DataAnalysis/TCGAmelanoma/Methylation/meTIL_probes.csv", row.name=1)
>>>>>>> 4203343b6214a072c3feb85611f00ef04d28d99e

dim(meTIL_probes)
meTIL_probe_info <- meTIL_probes[,1:3] # separating out the probe info from the probe values
meTIL_probes <- meTIL_probes[,4:478]
```


Changing identifier names and removing duplicates as was done before. 
```{r}
rid = tolower(substr(colnames(meTIL_probes),1,12))
rid = gsub("-", ".", rid)

colnames(meTIL_probes) <- rid

table(colnames(rnaseqMel)%in%colnames(meTIL_probes)) 
# All of the RNA-seq patient identifiers are also in the methylation identifiers

which(duplicated(colnames(meTIL_probes))) # There are 5 duplicates
colnames(meTIL_probes)[c(36,37,315,316,323,324,387,388,414,415)]

duplicated_SampleNames <- colnames(meTIL_probes)[duplicated(colnames(meTIL_probes))]

meTIL_duplicated<- meTIL_probes[,colnames(meTIL_probes)%in%duplicated_SampleNames]
colnames(meTIL_duplicated)
par(mfrow=c(2,3))
plot(meTIL_duplicated[,1],meTIL_duplicated[,2])
plot(meTIL_duplicated[,3],meTIL_duplicated[,4])
plot(meTIL_duplicated[,5],meTIL_duplicated[,6])
plot(meTIL_duplicated[,7],meTIL_duplicated[,8])
plot(meTIL_duplicated[,9],meTIL_duplicated[,10])
```
There seems to be more variation in the methylation 450K data compared to the RNA-seq data within the duplicates. But I'm not sure which one to take so i will drop the second data.

```{r cache=TRUE}
meTIL_probes <- meTIL_probes[,!duplicated(colnames(meTIL_probes))] # dropping the duplicates

dim(meTIL_probes)
dim(rnaseqMel)

table(colnames(meTIL_probes)%in%colnames(rnaseqMel))
# Theres 1 extra sample in meTIL_probes which is not in rnaseqMel

meTIL_probes <- meTIL_probes[,colnames(meTIL_probes )%in%colnames(rnaseqMel)] 

table(colnames(meTIL_probes) == colnames(rnaseqMel)) # Everything is in the same order and matches. 
```

```{r eval=FALSE}
write.csv(meTIL_probe_info, file="meTIL_probe_info.csv")
write.csv(meTIL_probes, file="meTIL_probes.csv")
```


#References

