---
title: "TCGA RNAseq, me450k and clinical data download"
author: "Antonio Ahn"
date: "`r Sys.Date()`"
output:
  pdf_document:
    highlight: default
    toc: yes
  github_document:
    toc: yes
  html_document:
    df_print: paged
    toc: yes
bibliography: /Users/antonioahn/Dropbox/GitHub/Downloading-TCGA-data/TCGAdatadownload.bib
---

Samur2014: RTCGAToolbox: a new tool for exporting TCGA Firehose data

Jeschke2017: DNA methylation-based immune response signature improves patient diagnosis in multiple cancers

Ekmekcioglu2016: Inflammatory Marker Testing Identifies CD74 Expression in Melanoma Tumor Cells, and Its Expression Associates with Favorable Survival for Stage III Melanoma


#Introduction

This document outlines the approach i took to download data level 3 TCGA data. RNA-seq, methylation 450K data and clinical information was downloaded using the `RTCGAToolbox`[@Samur2014] package and then processed (cleaned-up) for downstream analysis.

The `RTCGAToolbox` package written by Samur, retreives data from the broads institutive firehose database. Detailed information on how to use this package are available in the [Vignette](https://www.bioconductor.org/packages/devel/bioc/vignettes/RTCGAToolbox/inst/doc/RTCGAToolbox-vignette.html), online courses [PH525Xseries](http://genomicsclass.github.io/book/pages/tcga.html) and [Youtube](http://rafalab.github.io/pages/harvardx.html) videos. Many of the codes used here were taken from these resources. 

```{r message=FALSE}
#Load the library
library(RTCGAToolbox)
```

A list of available cancer types are found using the `getFirehoseDatasets` command. The run dates and analyze datas are found using the `getFirehoseRunningDates` and the `getFirehoseAnalyzeDates` commands. 

```{r}
getFirehoseDatasets()
head(getFirehoseRunningDates())
head(getFirehoseAnalyzeDates())
```
 
#Clinical and RNA-seq data - download and process

The clinical information and normalised RNA-seq data is downloaded.
The methylation data is downloaded separately from the clinical and RNA-seq data because its size was too large for my computer to handle. Thus the methylation data was downloaded in separately in our University departments computer cluster. 

Skin cutaneous melanoma (SKCM) is selected and "20160128" is used as the rundate. The default file size is 500 mb and this limit is extended to 3000 mb using `fileSizeLimit`.

The RNA-seq data (RNASeq2GeneNorm) contains normalised gene expression levels generated using **MapSplice** for alignment and **RNA-Seq by Expectation-Maximization (RSEM)** for quantification. RSEM values are calculated using an algorithm that estimate abundances at the gene level to generate TPM (Transcripts Per Million) values. TPM is similar to FPKM and RPKM in that it accounts for total number of reads and gene length. However TPM has gained more popularity over recent years because it is easier to interpret and more stable when comparing between samples. The normalisation is done by dividing the TPM values by the 75th percentile (3rd quartile) and multiplication by 1000.

```{r eval=FALSE}
getFirehoseDatasets()
head(getFirehoseRunningDates())
head(getFirehoseAnalyzeDates())

readDataMel <- getFirehoseData (dataset="SKCM", 
                                runDate="20160128",
                                forceDownload = TRUE, 
                                clinical = TRUE, 
                                RNASeq2GeneNorm = TRUE, 
                                Methylation = FALSE, fileSizeLimit= 3000)
```

```{r}
load("~/Dropbox/GitHub/Downloading-TCGA-data/RDatafiles/TCGA_readDataSKCM.RData")
```

```{r message=FALSE}
#extract the data
clinSKCM <- getData(readDataMel, "clinical")
rnaseqSKCM <- getData(readDataMel, "RNASeq2GeneNorm")
```

The clinical and RNA-seq data are processed (or cleaned up) before any downstream analysis.

1) The TCGA barcodes are structured differently between the clinical and RNA-seq datasets and thus needs to be changed/matched. For example the first TCGA barcode in the RNAseq data is "TCGA-3N-A9WB-06A-11R-A38C-07" whereas in the clinical data it is "tcga.d3.a2je"

2) Some samples that have 2 RNA-seq data. Duplicate RNA-seq data are removed.

3) Some samples contain RNA-seq data but there is no clinical information. Only those samples that have both RNA-seq and clinical information are retained for downstream analysis. 
```{r}
#Changing patient identifier names
dim(clinSKCM)
head(clinSKCM)

dim(rnaseqSKCM)
rnaseqSKCM[1:5,1:5]

#The identifiers in the RNA-seq data are transformed to be the same as the ones in the clinical data. This includes substringing the first 12 characters and replacing "-" with "."

rid = tolower(substr(colnames(rnaseqSKCM),1,12))
rid = gsub("-", ".", rid)  
colnames(rnaseqSKCM) = rid 

length(intersect(rid,rownames(clinSKCM))) # 469 samples intersect between RNAseq and clinSKCM
```

```{r}
#Remove duplicated samples
#Samples with duplicated names are removed. The data between the replicates are very similar however looking into the original barcodes, some of the duplicates are primary or metastatic, or a normal solid tumour. Tumours that are labeled primary, normal solid tumour or additional metastatic are removed. 

duplicatedSamples <- which(duplicated(colnames(rnaseqSKCM))) # 4  duplicate samples

duplicatedSampleNames <- colnames(rnaseqSKCM)[duplicated(colnames(rnaseqSKCM))]

rnaseqMel_duplicated <- rnaseqSKCM[,colnames(rnaseqSKCM) %in% duplicatedSampleNames] #matrix of only those duplicated samples

colnames(rnaseqMel_duplicated)
```

```{r figure1, fig.cap="\\label{fig:figure1} The duplicate samples are plotted together to assess the correlation. They look similar and it is not obvious which duplicate to keep."}
par(mfrow=c(2,2))
plot(log2(rnaseqMel_duplicated[1001:2000,1:2]))
plot(log2(rnaseqMel_duplicated[1001:2000,3:4]))
plot(log2(rnaseqMel_duplicated[1001:2000,5:6]))
plot(log2(rnaseqMel_duplicated[1001:2000,7:8]))
```

The full TCGA barcodes names of these duplicate samples are investigated. Information on TCGA barcodes are given [here](https://wiki.nci.nih.gov/display/TCGA/TCGA+barcode) and information on the sample type (e.g. primary, metastatic, additional metastsatic) from the TCGA barcode is provided [here](https://gdc.cancer.gov/resources-tcga-users/tcga-code-tables/sample-type-codes). 
```{r}
index <- which(colnames(rnaseqSKCM) %in% duplicatedSampleNames)

rnaseqSKCM2 <- getData(readDataMel, "RNASeq2GeneNorm")
original_rnaseq_barcode <- colnames(rnaseqSKCM2)
original_rnaseq_barcode[index]
```

#[1] "TCGA-D3-A1QA-07A-11R-A37K-07" "TCGA-D3-A1QA-06A-11R-A18T-07"
#[3] "TCGA-ER-A19T-06A-11R-A18U-07" "TCGA-ER-A19T-01A-11R-A18T-07"
#[5] "TCGA-ER-A2NF-06A-11R-A18T-07" "TCGA-ER-A2NF-01A-11R-A18T-07"
#[7] "TCGA-GN-A4U8-11A-11R-A32P-07" "TCGA-GN-A4U8-06A-11R-A32P-07"

The sample types of these duplicate samples are either of primary solid (01) tumour, metastastic (06), additional metastastic (07) or solid tissue (11) normal. 

Here i will retain metastatic tumours and thus will remove the primary solid tumour, additional metastatic and solid tissue normal. 

1st duplicate samples: remove the additional metastatic
2nd duplicate samples: remove the primary tumour 
3rd duplicate samples: remove the primary
4th duplicate samples: remove the solid tissue normal

```{r}
remove_index <- c("TCGA-D3-A1QA-07A-11R-A37K-07", "TCGA-ER-A19T-01A-11R-A18T-07",  "TCGA-ER-A2NF-01A-11R-A18T-07", "TCGA-GN-A4U8-11A-11R-A32P-07") #excluding these 

remove_index <- which(original_rnaseq_barcode %in% remove_index)

rnaseqSKCM = rnaseqSKCM[,-remove_index] # getting rid of the duplicate

dim(rnaseqSKCM) # from 473 samples to 469

length(intersect(colnames(rnaseqSKCM),rownames(clinSKCM)))  #469 samples interect between rnaseqSKCM and clinSKCM
length(rownames(clinSKCM)) # there is 1 sample in clinMel which there is absent in rnaseqMel 

clinSKCM <- clinSKCM[intersect(colnames(rnaseqSKCM),rownames(clinSKCM)),]
dim(clinSKCM)

table(colnames(rnaseqSKCM)==rownames(clinSKCM)) # patient names are in the same order
```

# Survival data clean-up and analysis

## Survival analysis: background

To analyse overall survival, 3 variables in the `clinMel` data set is required, which are "vital_status", "days_to_death" and "days_to_last_followup".

Information is available in a googles forum page [here](https://groups.google.com/forum/#!topic/ucsc-cancer-genomics-browser/YvKnWZSsw1Q) 

```{r}
dim(clinSKCM)
str(clinSKCM[,c("vital_status","days_to_death","days_to_last_followup")])
clinSKCM[1:5,c("vital_status","days_to_death","days_to_last_followup")]
```

* **vital_status:** "1" means deceased and "0" means still alive. 
* **days_to_death:** With patients who are deacesed, the days_to_death variable gives the number of days before death.
* **days_to_last_followup:** With patients who are still alive, the days_to_last_followup variable gives the number of days before the last follow-up.


## Survival data: Exploratory analysis

In 460 out of 469 patients, the days_to_death and days_to_last_followup are mutually exlcusive; if theres an NA in days_to_death then there is a number to DaystoLastfollowup and vice versa. The remaining have NA for both days_to_death and days_to_last_followup. 
```{r}
table(!is.na(clinSKCM[,"days_to_death"]) & is.na(clinSKCM[,"days_to_last_followup"])) #There are 220 patients who does not contain (NA) days_to_last_followup however contains days_to_death

table(is.na(clinSKCM[,"days_to_death"]) & !is.na(clinSKCM[,"days_to_last_followup"])) #There are 240 patients that do not contain (NA) days_to_death but contain days_to_last_followup

table(is.na(clinSKCM$"days_to_death") & is.na(clinSKCM$"days_to_last_followup")) #there are 9 patinets that contain NA for both days_to_death and days_to_last_followup
```

However there are some patients with both days_to_last_followup and days_to_death (69 patients). Also there are patients with both of these variables as NA (9 patients).
```{r}
survivalVariables <- c("days_to_last_followup","vital_status","days_to_death")

index <- !is.na(clinSKCM$"days_to_death") & !is.na(clinSKCM$"days_to_last_followup")

 table(is.na(clinSKCM$"days_to_death") & is.na(clinSKCM$"days_to_last_followup"))

clinMel[index,survivalVariables]
dim(clinMel[index,survivalVariables])

index <- is.na(clinMel[,"days_to_death"]) & is.na(clinMel[,"days_to_last_followup"])
clinMel[index,survivalVariables]
dim(clinMel[index,survivalVariables])
```

There is also 1 patient with a negative days_to_last_followup. What does this mean?
```{r}
survivalVariables <- c("days_to_last_followup","vital_status","days_to_death")

index <- which(clinSKCM[,"days_to_death"] < 0 | clinSKCM[,"days_to_last_followup"] < 0)

clinSKCM[index,survivalVariables]
```

## Survival analysis: merge days_to_death and days_to_last_followup

Here i merge days_to_death and days_to_last_followup to create a new variable called new_death. Most are simple to handle because they are mutually exlcusive; if there's an NA in days_to_death then there is a number to days_to_last_followup and vice versa. 

DELETE: However, as shown above, some patients have values to both variables with different number of days which i am unsure what that means. Also some patients have an NA to both variables. 

Here i create a new variable called new_death in which:
* If patient has deceased (1 in vital status), the days_to_death is selected
* If patient is alive (0 in vital status), days_to_last_followup is selected

```{r cache=TRUE}
mergeOS <- ifelse(clinSKCM[,"vital_status"]==1, clinSKCM[,"days_to_death"], clinSKCM[,"days_to_last_followup"] )

str(mergeOS)
table(is.na(mergeOS))
clinSKCM$mergeOS <- as.numeric(mergeOS)
```


## Survival analysis: sanity check with t-stage

Here i perform a "sanity check" to see if the survival data makes sense. First i look at the prognosis according to the tumour stage. 

* t0 - patients without a known primary tumor.
* t1
* t2
* t3
* t4
* ti
* tx

```{r}
library(survival)
```

```{r}
ev <- as.numeric(clinSKCM$vital_status)
fut <-as.numeric(clinSKCM$mergeOS)
su = Surv(fut, ev)

# There are 15 different types of T-stage. Here i reduce this number to 7 different groups. 
table(clinSKCM$pathology_T_stage)
table(substr(clinSKCM$pathology_T_stage,1,2)) 

t_stage = factor(substr(clinSKCM$pathology_T_stage,1,2))

plot(survfit(su~t_stage),mark.time=TRUE, lwd=2, col=1:7, las=1, cex.axis=1.5)
mtext("Overall Survival", side=2, line=2.7, cex=1.5)
mtext("Time (Days)", side=1, line=2.8, cex=1.5)

ntab = table(t_stage)
ns = paste("[n=", ntab, "]", sep="")
legend(8000, .8, col=1:7, lwd=2, legend=paste(levels(t_stage), ns))
text(6000,1, paste("p.value = 5.25e-06"))

summary(coxph(su~t_stage))
survdiff(su~t_stage)
```

There is a significant statistical difference in overall survival between the different T stages. 


## Survival analysis: sanity check with CD74

Another "sanity check" is done but this time with using the RNAseq data. CD74 gene exprresion was found to be associated with good prognosis using SKCM TCGA data [@Ekmekcioglu2016]. 

```{r}
#Patient samples are split into high and low CD74 expression using the median as the cut-off
CD74 <- ifelse(rnaseqSKCM["CD74",] > median(rnaseqSKCM["CD74",]), 1, 0)
# higher than median is 1, lower than median is 0
CD74 <- as.factor(CD74)
table(CD74)

ev <- as.numeric(clinSKCM$vital_status)
fut <-as.numeric(clinSKCM$mergeOS)
su = Surv(fut, ev)

plot(survfit(su~CD74),mark.time=TRUE, lwd=2, col=c("black","red"), las=1, cex.axis=1.5)
mtext("Overall Survival", side=2, line=2.7, cex=1.5)
mtext("Time (Days)", side=1, line=2.8, cex=1.5)

ntab = table(CD74)
ns = paste("[n=", ntab, "]", sep="")
legend(8000, .8, col= c("black","red"), lwd=2, legend=paste(levels(CD74), ns))
text(4000,0.8, paste("p.value= 1.94e-06"))

survdiff(su~CD74, data=clinSKCM)
```

As found in the study, higher CD74 gene expression is associated with a better prognosis.


# Methylation 450K data - download and processing

The methylation 450K data-frame was too big (>6gb) to download or work with in my desktop. It has 485,577 rows and 478 columns with each value having many digits. Therefore I had to use our cluster network to download the data and then reduce the file size by lowering the number of decimal points for every beta-value. The size-reduced file was then moved to my desktop and loaded into `R`. 

```{r eval=FALSE}
#This was done in our DSM cluster
#downloading the methylation 450k data from TCGA melanoma samples
readDataSKCM_methylation <- getFirehoseData(dataset = "SKCM",
                               runDate = "20160128",
                               forceDownload = TRUE,
                               clinical = FALSE,
                               RNASeq2GeneNorm = FALSE,
                               Methylation = TRUE,
                               fileSizeLimit = 3000)

me450kSKCM = getData(readDataSKCM_methylation, "Methylation",1) 
```

```{r eval=FALSE}
probeinfo <- me450kSKCM[,1:3] 
me450kSKCM <- me450kSKCM[,-(1:3)]

me450kSKCM <- sapply(me450kSKCM, as.numeric)

save.image("methylation450k_20160128_raw.RData")

#scp /home/STUDENT/ahnje770/methylation450k_20160128_raw.RData  /mnt/hcs/dsm-eccles-seq


names(as.list(.GlobalEnv)) #too look at the varialbes in the global environment (in the DSM cluster)

#to load it in the future in DSM cluster
load("/mnt/hcs/dsm-eccles-seq/methylation450k_files/methylation450k_20160128_raw.RData")
```


##Change the identifier names in the methylation data

```{r eval=FALSE}
rid = tolower(substr(colnames(me450kSKCM),1,12))
rid = gsub("-", ".", rid)

duplicatedSampleNames_me450k <- rid[which(duplicated(rid))]
colnames(me450kSKCM)[rid %in% duplicatedSampleNames_me450k]
# [1] "TCGA-D3-A1QA-07A-11D-A373-05" "TCGA-D3-A1QA-06A-11D-A19B-05"
# [3] "TCGA-ER-A19T-06A-11D-A19D-05" "TCGA-ER-A19T-01A-11D-A19D-05"
# [5] "TCGA-ER-A2NF-06A-11D-A19D-05" "TCGA-ER-A2NF-01A-11D-A19D-05"
# [7] "TCGA-FW-A3R5-11A-11D-A23D-05" "TCGA-FW-A3R5-06A-11D-A23D-05"
# [9] "TCGA-GN-A4U8-11A-11D-A32S-05" "TCGA-GN-A4U8-06A-11D-A32S-05"

#Just as the RNA-seq data, here I remove the additional metastatic, primary tumours, and normal solid tissues 

#removing the duplicated samples here
#1st duplicate: remove the additional metastatic (07)
#2nd duplicate: remove the primary tumour (01)
#3rd duplicate: remove the primary (01)
#4th duplicate: remove the solid tissue normal (11)
#5th duplicate: remove the solid tissue normal (11)

#01 Primary solid tumour 
#06 metastatic
#07 additional metastastatic
#11 Solid tissue normal 

remove_index <- c("TCGA-D3-A1QA-07A-11D-A373-05", "TCGA-ER-A19T-01A-11D-A19D-05" ,  "TCGA-ER-A2NF-01A-11D-A19D-05", "TCGA-FW-A3R5-11A-11D-A23D-05", "TCGA-GN-A4U8-11A-11D-A32S-05")
remove_index <- which(colnames(me450kSKCM) %in% remove_index)
remove_index #the rows i will be removing for the duplicated samples
#[1]  36 316 324 387 414

me450kSKCM = me450kSKCM[,-remove_index] # getting rid of the duplicate

dim(me450kSKCM) # from 475 samples to 470
table(duplicated(colnames(me450kSKCM)))
#FALSE 
#  470

rid = tolower(substr(colnames(me450kSKCM),1,12))
rid = gsub("-", ".", rid)

colnames(me450kSKCM) <- rid

length(intersect(colnames(me450kSKCM),colnames(rnaseqSKCM)))  #469 samples interect between meTIL_probes and rnaseqSKCM
length(colnames(meTIL_probes)) # there is 1 sample in clinMel which there is absent in rnaseqMel 

me450kSKCM <- me450kSKCM[,intersect(colnames(me450kSKCM),colnames(rnaseqSKCM))]
dim(meTIL_probes)

table(colnames(rnaseqSKCM)==colnames(me450kSKCM)) # patient names are in the same order
```


##Reducing the size of the methylation 450K data

```{r eval=FALSE}
str(me450kSKCM) # this shows that all the values are characters. 

me450kSKCM <- sapply(me450kSKCM, as.numeric) 
me450kSKCM_rounded <- as.matrix(round(me450kSKCM, digits=3)) # Round to 3 digits
write.csv(me450kSKCM_rounded, "me450kSKCM_rounded.csv")

#scp /home/STUDENT/ahnje770/me450kSKCM_rounded.csv /mnt/hcs/dsm-eccles-seq

rm(me450kSKCM) #remove the unrounded me450kfile
setwd("/mnt/hcs/dsm-eccles-seq/methylation450k_files")

save.image("methylation450k_20160128_rounded.RData")
```

After i reduced the size of the methylation data to generate `me450kMel_rounded`, I saved into my computer for loading. 
```{r}
load("/Volumes/dsm-eccles-seq/methylation450k_files/methylation450k_20160128_rounded.RData") #"/Volumes" (on my dekstop) is the same as "/mnt/hcs" (in DSM cluster)

dim(me450kSKCM_rounded)
dim(probeinfo)
class(me450kSKCM_rounded)

me450kSKCM_rounded[1:3,1:3]
```

##Acquring methylation probe values for meTIL-score

It was demonstrated that methylation probe values can be used to determine the level of CD8 immune cells within bulk tumour [@ref3]. 

Beta-values of 5 CpG probes are needed to generate the meTIL-score. Here i did not use me450kMel_rounded but used the data prior to rounding to 3 decimal points.

```{r eval=FALSE}
meTIL_probes <- c("cg20792833","cg20425130","cg23642747","cg12069309","cg21554552") # the 5 CpG probes needed to generate the meTIL-score.

me450kMel[1:3,1:6] 
```
            X Gene_Symbol Chromosome Genomic_Coordinate
1 cg00000029        RBL2         16           53468112
2 cg00000108     C3orf35          3           37459206
3 cg00000109      FNDC3B          3          171916037
  TCGA.3N.A9WB.06A.11D.A38H.05 TCGA.3N.A9WC.06A.11D.A38H.05
1                       0.5167                       0.4193
2                           NA                           NA
3                           NA                           NA
```{r eval=FALSE}
write.csv(me450kMel[me450kMel$X%in%probes_iwant,], file="meTIL_probes.csv")
```

The "meTIL_probes.csv" file is transfered from the server to my computer and then loaded. 

```{r cache=TRUE}
meTIL_probes <- read.csv("~/Dropbox/Education/Bioinformatics/5DataAnalysis/TCGAmelanoma/Methylation/meTIL_probes.csv", row.name=1)
meTIL_probes <- read.csv("~/Dropbox/Education/Bioinformatics/5. DataAnalysis/TCGAmelanoma/Methylation/meTIL_probes.csv", row.name=1)

dim(meTIL_probes)
meTIL_probe_info <- meTIL_probes[,1:3] # separating out the probe info from the probe values
meTIL_probes <- meTIL_probes[,4:478]
```


Changing identifier names and removing duplicates as was done before. 
```{r}
rid = tolower(substr(colnames(meTIL_probes),1,12))
rid = gsub("-", ".", rid)

colnames(meTIL_probes) <- rid

table(colnames(rnaseqMel)%in%colnames(meTIL_probes)) 
# All of the RNA-seq patient identifiers are also in the methylation identifiers

which(duplicated(colnames(meTIL_probes))) # There are 5 duplicates
colnames(meTIL_probes)[c(36,37,315,316,323,324,387,388,414,415)]

duplicated_SampleNames <- colnames(meTIL_probes)[duplicated(colnames(meTIL_probes))]

meTIL_duplicated<- meTIL_probes[,colnames(meTIL_probes)%in%duplicated_SampleNames]
colnames(meTIL_duplicated)
par(mfrow=c(2,3))
plot(meTIL_duplicated[,1],meTIL_duplicated[,2])
plot(meTIL_duplicated[,3],meTIL_duplicated[,4])
plot(meTIL_duplicated[,5],meTIL_duplicated[,6])
plot(meTIL_duplicated[,7],meTIL_duplicated[,8])
plot(meTIL_duplicated[,9],meTIL_duplicated[,10])
```
There seems to be more variation in the methylation 450K data compared to the RNA-seq data within the duplicates. But I'm not sure which one to take so i will drop the second data.

```{r cache=TRUE}
meTIL_probes <- meTIL_probes[,!duplicated(colnames(meTIL_probes))] # dropping the duplicates

dim(meTIL_probes)
dim(rnaseqMel)

table(colnames(meTIL_probes)%in%colnames(rnaseqMel))
# Theres 1 extra sample in meTIL_probes which is not in rnaseqMel

meTIL_probes <- meTIL_probes[,colnames(meTIL_probes )%in%colnames(rnaseqMel)] 

table(colnames(meTIL_probes) == colnames(rnaseqMel)) # Everything is in the same order and matches. 
```

```{r eval=FALSE}
write.csv(meTIL_probe_info, file="meTIL_probe_info.csv")
write.csv(meTIL_probes, file="meTIL_probes.csv")
```


#References

